In [2]:
%load_ext autoreload
%autoreload 2

In [6]:
from sklearn.ensemble import RandomForestClassifier

from src.hyperparameters import hyperparams_rf
from src.model_runner import ModelRunner, PredictionMetrics
from src.models import RandomForest as MyRandomForest, Model

runner = ModelRunner('dataset/Occupancy_Estimation.csv')

(1000, 18) (1000,)


In [8]:
hp_cfg, metrics = runner.run(MyRandomForest, hyperparams_rf, PredictionMetrics.classification_metrics())

(1000, 18) (1000,)
Hyperparams = {'method': 'gini', 'n_estimators': 2, 'max_depth': 3, 'max_features': 5}
Fold 0
{'accuracy': np.float64(0.95), 'precision': np.float64(0.896551724137931), 'recall': np.float64(0.9285714285382654), 'f_measure': np.float64(0.912280701738381), 'exec_time': 0.05271196365356445}
Fold 1
{'accuracy': np.float64(0.95), 'precision': np.float64(0.9642857142857143), 'recall': np.float64(0.8709677419073881), 'f_measure': np.float64(0.9152542372726229), 'exec_time': 0.07412409782409668}
Fold 2
{'accuracy': np.float64(0.88), 'precision': np.float64(0.967741935483871), 'recall': np.float64(0.7317073170553243), 'f_measure': np.float64(0.8333333333217592), 'exec_time': 0.038186073303222656}
Fold 3
{'accuracy': np.float64(0.87), 'precision': np.float64(0.875), 'recall': np.float64(0.813953488353164), 'f_measure': np.float64(0.8433734939657425), 'exec_time': 0.025348901748657227}
Fold 4
{'accuracy': np.float64(0.99), 'precision': np.float64(1.0), 'recall': np.float64(0.97

In [9]:
print("Optimized hyperparameters:\n", hp_cfg)
print("Metrics:\n", metrics)

Optimized hyperparameters:
 {'method': 'gini', 'n_estimators': 20, 'max_depth': 5, 'max_features': 5}
Metrics:
 {'accuracy': { mean=0.998, std=0.0040000000000000036, conf_interval=(np.float64(0.9949837904495279), np.float64(1.0010162095504722)) }, 'precision': { mean=0.9942434210526315, std=0.011565903612992867, conf_interval=(np.float64(0.9855221238182942), np.float64(1.0029647182869688)) }, 'recall': { mean=0.9999999999714664, std=4.519032180514679e-12, conf_interval=(np.float64(0.9999999999680588), np.float64(0.999999999974874)) }, 'f_measure': { mean=0.9970793650651848, std=0.005868810163098502, conf_interval=(np.float64(0.9926539747492233), np.float64(1.0015047553811462)) }, 'exec_time': { mean=1.3117542505264281, std=0.2634625058216569, conf_interval=(np.float64(1.1130897189637778), np.float64(1.5104187820890784)) }}


In [13]:
hp_cfg = {'criterion': 'gini', 'n_estimators': 20, 'max_depth': 5, 'max_features': 5}

Optimized hyperparameters:
 {'method': 'gini', 'n_estimators': 20, 'max_depth': 5, 'max_features': 5}
Metrics:
 {'accuracy': { mean=0.998, std=0.0040000000000000036, conf_interval=(np.float64(0.9949837904495279), np.float64(1.0010162095504722)) }, 'precision': { mean=0.9942434210526315, std=0.011565903612992867, conf_interval=(np.float64(0.9855221238182942), np.float64(1.0029647182869688)) }, 'recall': { mean=0.9999999999714664, std=4.519032180514679e-12, conf_interval=(np.float64(0.9999999999680588), np.float64(0.999999999974874)) }, 'f_measure': { mean=0.9970793650651848, std=0.005868810163098502, conf_interval=(np.float64(0.9926539747492233), np.float64(1.0015047553811462)) }, 'exec_time': { mean=1.3117542505264281, std=0.2634625058216569, conf_interval=(np.float64(1.1130897189637778), np.float64(1.5104187820890784)) }}

In [14]:
def pred_caller(rf: Model):
    def f(x_test):                
        return rf.predict(x_test)
    return f

runner.shap(MyRandomForest, hp_cfg, pred_caller=pred_caller)

ValueError: Image size of 694513x600 pixels is too large. It must be less than 2^16 in each direction.

<Figure size 800x650 with 3 Axes>

In [19]:
library_hp_cfg, library_metrics = runner.run(RandomForestClassifier, hyperparams_rf, PredictionMetrics.classification_metrics())

Hyperparams = {'criterion': 'gini', 'n_estimators': 2, 'max_depth': 3, 'max_features': 5}
Fold 0
{'accuracy': np.float64(1.0), 'precision': np.float64(1.0), 'recall': np.float64(0.9999999999642857), 'f_measure': np.float64(0.999999999982143), 'exec_time': 0.003757953643798828}
Fold 1
{'accuracy': np.float64(0.98), 'precision': np.float64(0.967741935483871), 'recall': np.float64(0.9677419354526535), 'f_measure': np.float64(0.9677419354682623), 'exec_time': 0.002150297164916992}
Fold 2
{'accuracy': np.float64(0.99), 'precision': np.float64(1.0), 'recall': np.float64(0.9756097560737657), 'f_measure': np.float64(0.987654320975461), 'exec_time': 0.0020742416381835938}
Fold 3
{'accuracy': np.float64(0.98), 'precision': np.float64(0.9555555555555556), 'recall': np.float64(0.9999999999767443), 'f_measure': np.float64(0.9772727272616218), 'exec_time': 0.0018148422241210938}
Fold 4
{'accuracy': np.float64(0.99), 'precision': np.float64(0.9772727272727273), 'recall': np.float64(0.9999999999767443

In [21]:
print("Optimized hyperparameters:\n", library_hp_cfg)
print("Metrics:\n", library_metrics)

Optimized hyperparameters:
 {'criterion': 'entropy', 'n_estimators': 20, 'max_depth': 3, 'max_features': 5}
Metrics:
 {'accuracy': { mean=0.9970000000000001, std=0.004582575694955844, conf_interval=(np.float64(0.9935444978557783), np.float64(1.0004555021442219)) }, 'precision': { mean=0.9915407183499289, std=0.012976706073540446, conf_interval=(np.float64(0.9817556021517583), np.float64(1.0013258345480995)) }, 'recall': { mean=0.9999999999714664, std=4.519032180514679e-12, conf_interval=(np.float64(0.9999999999680588), np.float64(0.999999999974874)) }, 'f_measure': { mean=0.995709502051524, std=0.00658254626319372, conf_interval=(np.float64(0.9907459173251566), np.float64(1.0006730867778915)) }, 'exec_time': { mean=0.01123950481414795, std=0.00025930213177799254, conf_interval=(np.float64(0.011043977422566308), np.float64(0.01143503220572959)) }}


Optimized hyperparameters:
 {'criterion': 'entropy', 'n_estimators': 20, 'max_depth': 3, 'max_features': 5}
Metrics:
 {'accuracy': { mean=0.9970000000000001, std=0.004582575694955844, conf_interval=(np.float64(0.9935444978557783), np.float64(1.0004555021442219)) }, 'precision': { mean=0.9915407183499289, std=0.012976706073540446, conf_interval=(np.float64(0.9817556021517583), np.float64(1.0013258345480995)) }, 'recall': { mean=0.9999999999714664, std=4.519032180514679e-12, conf_interval=(np.float64(0.9999999999680588), np.float64(0.999999999974874)) }, 'f_measure': { mean=0.995709502051524, std=0.00658254626319372, conf_interval=(np.float64(0.9907459173251566), np.float64(1.0006730867778915)) }, 'exec_time': { mean=0.01123950481414795, std=0.00025930213177799254, conf_interval=(np.float64(0.011043977422566308), np.float64(0.01143503220572959)) }}

In [23]:
runner.shap(RandomForestClassifier, library_hp_cfg, pred_caller=pred_caller, show=False)

ValueError: Image size of 694513x600 pixels is too large. It must be less than 2^16 in each direction.

<Figure size 800x650 with 9 Axes>